File: 프로그래머스/1/42889. 실패율/실패율.py
from collections import defaultdict
def solution(N, stages):
    answer = []
    cnt = defaultdict(int)
    
    for stage in stages:
        cnt[stage] += 1
    
    result = []
    for i in range(1, N+1):
        #스테이지에 도달했으나 아직 클리어하지 못한 플레이어의 수
        yet = cnt[i]
        #스테이지에 도달한 플레이어 수
        complete = 0
        for j in range(i, N+2):
            complete += cnt[j]
        if complete == 0:
            result.append((i, 0))
        else:
            result.append((i, yet/complete))
    
    result.sort(key=lambda x: (-x[1], x[0]))
    return [x[0] for x in result]

File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.*;
class Solution {
    public int[] solution(int N, int[] stages) {
        List<double[]> answer = new ArrayList<>();
        HashMap<Integer, Integer> people = new HashMap<>();
        
        for(int stage: stages) {
            people.putIfAbsent(stage, 0);
            people.put(stage, people.get(stage)+1);
        }
        
        int cnt = stages.length;
        for(int i=1; i<N+1; i++) {
            if(!people.containsKey(i)) {
                answer.add(new double[]{i, 0});
            } else {
                
                answer.add(new double[]{i, (double) people.get(i)/cnt});
                cnt -= people.get(i);
            }
        }
        
        answer.sort((a, b) -> {
            if(b[1] == a[1]) {
                return (int) (a[0] - b[0]);
            }
            return Double.compare(b[1], a[1]);
        });
        
        int[] result = new int[N];
        int i=0;
        for(double[] a : answer) {
            result[i] = (int) a[0];
            i++;
        }
        

        return result;
    }
}